pkg/monitor/resourcewatch/storage/monitor_store.go

```diff
@@ -0,0 +1,295 @@
+package storage
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/origin/pkg/monitor"
+	"github.com/openshift/origin/pkg/monitor/monitorapi"
+
+	corev1 "k8s.io/api/core/v1"
+	_ "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+)
+
+var (
+	coreScheme = runtime.NewScheme()
+	coreCodecs = serializer.NewCodecFactory(coreScheme)
+)
+
+type monitorStorage struct {
+	monitor     *monitor.Monitor
+	artifactDir string
+	client      kubernetes.Interface
+	startTime   time.Time
+}
+
+// NewMonitorStorage creates a monitor store for resource watch
+func NewMonitorStorage(artifactDir string, client kubernetes.Interface) (ResourceWatchStore, error) {
+	storage := &monitorStorage{
+		monitor:     monitor.NewMonitorWithInterval(time.Hour),
+		artifactDir: artifactDir,
+		client:      client,
+		startTime:   time.Now(),
+	}
+
+	return storage, nil
+}
+
+func decodeObject(objUnstructured *unstructured.Unstructured, obj runtime.Object) error {
+	objectBytes, err := runtime.Encode(unstructured.UnstructuredJSONScheme, objUnstructured)
+	if err != nil {
+		return fmt.Errorf("encoding unstructured error: %v", err)
+	}
+	err = runtime.DecodeInto(coreCodecs.UniversalDecoder(corev1.SchemeGroupVersion), objectBytes, obj)
+	if err != nil {
+		return fmt.Errorf("decoding object error: %v", err)
+	}
+	return nil
+}
+
+func (s *monitorStorage) handleEvents(objUnstructured *unstructured.Unstructured) {
+	eventObj := corev1.Event{}
+	err := decodeObject(objUnstructured, &eventObj)
+	if err != nil {
+		klog.Warningf("Decoding %s failed with error: %v", objUnstructured.GetName(), err)
+	}
+	monitor.EventAddOrUpdateFunc(context.TODO(), s.monitor, s.client, s.startTime, &eventObj)
+}
+
+// OnAdd processes add event for all monitored resources
+func (s *monitorStorage) OnAdd(obj interface{}) {
+	objUnstructured, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		klog.Warningf("Object is not unstructured: %v", obj)
+	}
+	gvk := objUnstructured.GroupVersionKind()
+	switch {
+	case gvk.Group == "" && gvk.Kind == "Node":
+		{
+			nodeObj := corev1.Node{}
+			err := decodeObject(objUnstructured, &nodeObj)
+			if err != nil {
+				klog.Warningf("Decoding %s failed with error: %v", objUnstructured.GetName(), err)
+			}
+			monitor.NodeAddFunc(nodeObj)
```

Comment:
This will get called when we fail to decode, looks like we should return here and for all examples like this below.